File: java/BasicParser.java
import java.util.HashSet;
import java.util.List;
import java.io.StringReader;;

public class BasicParser {
    HashSet<String> reserved = new HashSet<String>();
    Parser.Operators operators = new Parser.Operators();
    Parser expr0 = Parser.rule();
    Parser primary = Parser.rule(PrimaryExpr.class)
        .or(Parser.rule().sep("(").ast(expr0).sep(")"),
            Parser.rule().number(NumberLiteral.class),
            Parser.rule().identifier(Name.class, reserved),
            Parser.rule().string(StringLiteral.class));
    Parser factor = Parser.rule().or(Parser.rule(NegativeExpr.class).sep("-").ast(primary),
                              primary);                               
    Parser expr = expr0.expression(BinaryExpr.class, factor, operators);

    Parser statement0 = Parser.rule();
    Parser block = Parser.rule(BlockStmnt.class)
        .sep("{").option(statement0)
        .repeat(Parser.rule().sep(";", Token.EOL).option(statement0))
        .sep("}");
    Parser simple = Parser.rule(PrimaryExpr.class).ast(expr);
    Parser statement = statement0.or(
            Parser.rule(IfStmnt.class).sep("if").ast(expr).ast(block)
                               .option(Parser.rule().sep("else").ast(block)),
            Parser.rule(WhileStmnt.class).sep("while").ast(expr).ast(block),
            simple);

    Parser program = Parser.rule().or(statement, Parser.rule(NullStmnt.class))
                           .sep(";", Token.EOL);

    public BasicParser() {
        reserved.add(";");
        reserved.add("}");
        reserved.add(Token.EOL);

        operators.add("=", 1, Parser.Operators.RIGHT);
        operators.add("==", 2, Parser.Operators.LEFT);
        operators.add(">", 2, Parser.Operators.LEFT);
        operators.add("<", 2, Parser.Operators.LEFT);
        operators.add("+", 3, Parser.Operators.LEFT);
        operators.add("-", 3, Parser.Operators.LEFT);
        operators.add("*", 4, Parser.Operators.LEFT);
        operators.add("/", 4, Parser.Operators.LEFT);
        operators.add("%", 4, Parser.Operators.LEFT);
    }
    public ASTree parse(Lexer lexer) throws ParseException {
        return program.parse(lexer);
    }
    
    public static void main(String[] args) throws ParseException {
        StringReader stringReader = new StringReader(BasicTest.lexer);
        Lexer l = new Lexer(stringReader);
        BasicParser bp = new BasicParser();
        while (l.peek(0) != Token.EOF) {
            ASTree ast = bp.parse(l);
            System.out.println("=> " + ast.toString());
        }
    }
}

File: parser/BasicParser.ts
import { Precedence, Parser } from './Parser'
import {
    PrimaryExpr,
    NumberLiteral,
    Name,
    NegativeExpr,
    BinaryExpr,
    NullStmnt,
    StringLiteral,
    BlockStmnt,
    IfStmnt,
    WhileStmnt,
    ASTree,
} from '../ast/ASTree'
import { Token } from '../ast/Token'
import Lexer from '../ast/Lexer'
import { BasicEnv } from '../eval/Environment'

export default class BasicParser {
    reserved = [';', '}', Token.EOL]
    operators = {
        '=': new Precedence(1, false),
        '==': new Precedence(2),
        '!=': new Precedence(2),
        '<': new Precedence(2),
        '<=': new Precedence(2),
        '>': new Precedence(2),
        '>=': new Precedence(2),
        '+': new Precedence(3),
        '-': new Precedence(3),
        '*': new Precedence(4),
        '/': new Precedence(4),
        '%': new Precedence(4),
    }

    expr0 = Parser.rule()
    primary = Parser.rule(PrimaryExpr).or(
        Parser.rule().sep('(').ast(this.expr0).sep(')'),
        Parser.rule().number(NumberLiteral),
        Parser.rule().identifier(Name, this.reserved),
        Parser.rule().string(StringLiteral)
    )
    factor = Parser.rule('factor').or(
        Parser.rule(NegativeExpr).sep('-').ast(this.primary),
        this.primary
    )
    expr = this.expr0.expression(BinaryExpr, this.factor, this.operators)

    statement0 = Parser.rule()
    block = Parser.rule(BlockStmnt)
        .sep('{')
        .option(this.statement0)
        .repeat(Parser.rule().sep(';', Token.EOL).option(this.statement0))
        .sep('}')
    simple = Parser.rule(PrimaryExpr).ast(this.expr)
    statement = this.statement0.or(
        Parser.rule(IfStmnt)
            .sep('if')
            .ast(this.expr)
            .ast(this.block)
            .option(Parser.rule().sep('else').ast(this.block)),
        Parser.rule(WhileStmnt).sep('while').ast(this.expr).ast(this.block),
        this.simple
    )

    program = Parser.rule('program')
        .or(this.statement, Parser.rule(NullStmnt))
        .sep(';', Token.EOL)

    constructor() {}

    parse(lexer: Lexer): ASTree {
        return this.program.parse(lexer)
    }
}

// const lexer = new Lexer('./test-lexer.stone')
// lexer.process().then(() => {
//     const env = new BasicEnv()
//     const basicParser = new BasicParser()
//     while (lexer.peek(0) != Token.EOF) {
//         const ast = basicParser.parse(lexer)
//         console.log('=> ', ast.toString())
//         if (!(ast instanceof NullStmnt)) {
//             console.log('eval =>', ast.eval(env))
//         }
//     }
// })
